crates/lib3h/src/engine/ghost_engine.rs

```diff
@@ -549,65 +552,90 @@ impl<'engine> GhostEngine<'engine> {
+        space_address: Address,
+        from_agent_id: Address,
+        to_agent_id: Address,
+        net_msg: P2pProtocol,
+    ) -> Lib3hResult<(
+        &mut GatewayParentWrapper<GhostEngine<'engine>, P2pGateway>,
+        Opaque,
+    )> {
+        let chain_id = (space_address, from_agent_id);
 
         let maybe_this_peer = self.this_space_peer(chain_id.clone());
         if let Err(error) = maybe_this_peer {
-            ghost_message.respond(Err(error))?;
-            return Ok(());
+            return Err(error);
         };
         let this_peer = maybe_this_peer.unwrap();
 
-        let to_agent_id: String = msg.to_agent_id.clone().into();
-        if &this_peer.peer_address == &to_agent_id {
+        if &this_peer.peer_address == &to_agent_id.to_string() {
             return Err(Lib3hError::new_other("messaging self not allowed"));
         }
 
-        let net_msg = P2pProtocol::DirectMessage(msg.clone());
-
         // Serialize payload
         let mut payload = Vec::new();
         net_msg
             .serialize(&mut Serializer::new(&mut payload))
             .unwrap();
-        // Send
-        let peer_address: String = msg.to_agent_id.clone().into();
 
         let space_gateway = self
             .space_gateway_map
             .get_mut(&chain_id)
             .ok_or_else(|| Lib3hError::new_other("Not part of that space"))?;
 
-        space_gateway
-            .request(
-                span,
-                GatewayRequestToChild::Transport(
-                    transport::protocol::RequestToChild::SendMessage {
-                        uri: Url::parse(&("agentId:".to_string() + &peer_address))
-                            .expect("invalid url format"),
-                        payload: payload.into(),
-                    },
-                ),
-                Box::new(|_me, response| {
-                    debug!("GhostEngine: response to handle_direct_message message: {:?}", response);
-                    match response {
-                        GhostCallbackData::Response(Ok(
-                                GatewayRequestToChildResponse::Transport(
-                                    transport::protocol::RequestToChildResponse::SendMessageSuccess
-                        ))) => {
-                            panic!("We Need to bookmark this ghost_message so that we can invoke it with a message from the remote peer when they send it back");
-                        }
-                        _ => ghost_message.respond(Err(format!("{:?}", response).into()))?,
-                    };
-                    Ok(())
-                })
-            )?;
+        Ok((space_gateway.as_mut(), Opaque::from(payload)))
+    }
+
+    fn handle_direct_message(
+        &mut self,
+        span: Span,
+        ghost_message: ClientToLib3hMessage,
+        mut msg: DirectMessageData,
+    ) -> Lib3hResult<()> {
+        let to_agent_id = msg.to_agent_id.clone();
+
+        // Generate a new request_id
```

Comment:
```suggestion
        // Generate a new request_id for the network transport exchange.
        // we can overwrite the value in the DirectMessageData because the ghost tracker will handle
        // the request response pairing 
```